File: solutions/elixir/luhn/1/lib/luhn.ex
defmodule Luhn do
  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t()) :: boolean
  def valid?("0"), do: false
  def valid?(" 0"), do: false
  def valid?(number) do
    number = String.replace(number, " ", "")
    if String.match?(number, ~r/^[[:digit:]]+$/) do
      luhn = number
      |> String.reverse()
      |> String.graphemes()
      |> Enum.map(&String.to_integer/1)
      |> Enum.with_index()
      |> Enum.map(fn {digit, i} ->
        if rem(i, 2) == 1 do
          digit = digit * 2
          if digit > 9, do: digit - 9, else: digit
        else
          digit
        end
      end)
      |> Enum.sum()

      rem(luhn, 10) == 0
    else
      false
    end
  end
end

File: solutions/ruby/luhn/1/luhn.rb

class Luhn
    def self.valid?(input)
        input = input.gsub(/\s+/, "")
        return false if input.length <= 1 || input.match?(/\D/)

        (input
            .chars.map(&:to_i)
            .reverse.map.with_index do |num, index|
                if index.odd?
                    num *= 2
                    num -= 9 if num >= 10
                end
                num
            end
            .sum % 10).zero?
    end
end